source/Patches/RainbowMod/PlayerTabPatch.cs

```diff
@@ -1,39 +1,76 @@
+        [HarmonyPatch(nameof(PlayerTab.OnEnable))]
+        public static void OnEnablePatch(PlayerTab __instance)
         {
-            PlayerControl.SetPlayerMaterialColors(PlayerControl.LocalPlayer.Data.ColorId, __instance.DemoImage);
-            __instance.HatImage.SetHat(SaveManager.LastHat, PlayerControl.LocalPlayer.Data.ColorId);
-            PlayerControl.SetSkinImage(SaveManager.LastSkin, __instance.SkinImage);
-            PlayerControl.SetPetImage(SaveManager.LastPet, PlayerControl.LocalPlayer.Data.ColorId, __instance.PetImage);
+            foreach (ColorChip instanceColorChip in __instance.ColorChips)
+                instanceColorChip.gameObject.Destroy();
+
+            __instance.ColorChips.Clear();
+
             var colors = Palette.PlayerColors;
             var num = colors.Length / 4f;
-            for (int i = 0;i < colors.Length;i++)
+
+            for (int i = 0; i < colors.Length; i++)
             {
-                var x = __instance.XRange.Lerp((i % 4) / 4f) + 0.25f;
-                var y = __instance.YStart - (i / 4) * 0.55f;
+                var x = __instance.XRange.Lerp((i % 4) / 4f) + 0.26f;
+                var y = (__instance.YStart - 0.16f) - (i / 4) * 0.5f;
                 var colorChip = Object.Instantiate(__instance.ColorTabPrefab, __instance.ColorTabArea, true);
                 colorChip.transform.localScale *= 0.8f;
                 colorChip.transform.localPosition = new Vector3(x, y, -1f);
                 var colorId = (byte)i;
-                colorChip.Button.OnClick.AddListener((Action) (() =>
+
+                colorChip.Button.OnClick.AddListener((Action)(() =>
                 {
-                    __instance.SelectColor(colorId);
-                    if (colorId <= 17) SaveManager.BodyColor = colorId;
+                    if (colorId <= 17) __instance.SelectColor(colorId);
+                    __instance.SelectCustomColor(colorId);
+                    __instance.HatImage.SetColor(colorId);
                 }));
+
                 colorChip.Inner.color = colors[i];
                 __instance.ColorChips.Add(colorChip);
             }
-            return false;
+        }
+
+        [HarmonyPostfix]
+        [HarmonyPatch(nameof(PlayerTab.Update))]
+        public static void UpdatePatch(PlayerTab __instance)
+        {
+            for (int i = 0; i < __instance.ColorChips.Count; i++)
+            {
+                if (RainbowUtils.IsRainbow(i))
+                {
+                    __instance.ColorChips[i].Inner.color = RainbowUtils.Rainbow;
+                    break;
+                }
+            }
+        }
+
+        public static void SelectCustomColor(this PlayerTab playerTab, byte colorId)
+        {
+            PluginSingleton<TownOfUs>.Instance.CustomColor.Value = colorId;
+            if (colorId <= 17) return;
+
+            playerTab.UpdateAvailableColors();
+
+            if (PlayerControl.LocalPlayer)
+                PlayerControl.LocalPlayer.CmdCheckColor(colorId);
```

Comment:
```suggestion
            PlayerControl.LocalPlayer?.CmdCheckColor(colorId);
```
although its probably always present here so 
```suggestion
            PlayerControl.LocalPlayer.CmdCheckColor(colorId);
```
is also valid